### PR TITLE
adding prefetching_iterator and related tests used for prefetching containers within lambda functions

### PR DESCRIPTION
**Pull request by zkhatami88**
prefetching_iterator is created within loop.hpp to prefetch all containers within lambda functions.
for_each.hpp is changed as well to work with loop_n within loop.hpp to call prefetching_iterator.
foreach_prefetching.cpp, foreach_prefetching_executors.cpp and foreach_tests_prefetching.hpp are related tests created for testing the prefetching_iterator performance.

### CODE REVIEW

**Comment by sithhell** [comment]
The files you added miss a newline at the end. This needs to get fixed.
In addition, would it be possible to have the prefetch stuff in a seperate header instead of having it all in the for_loop header? Could we reuse it for other algorithms as well?

File: hpx/parallel/util/loop.hpp [hunk 1]
```diff
@@ -68,12 +69,176 @@ namespace hpx { namespace parallel { namespace util
...
+
+            explicit prefetcher_context (std::size_t begin, std::size_t end,
+                std::size_t p_factor, std::initializer_list< T * > &&l)
+            {
+                std::size_t vector_size = end - begin + 1;
+                range.resize(vector_size);
+                for(std::size_t i=0; i<vector_size; ++i) range[i] = i + begin;
+                it_begin = range.begin();
+                it_end = range.begin() + vector_size - 1;
+                chunk_size = p_factor * 64ul / sizeof(T);
```
Comment (sithhell) [inline hpx/parallel/util/loop.hpp]: where does the 64 come from?

File: hpx/parallel/util/loop.hpp [hunk 2]
```diff
@@ -96,23 +261,88 @@ namespace hpx { namespace parallel { namespace util
...
+
+                    for(/**/; j < std::min( (it.range_size ),(it.idx + it.chunk_size) ); ++j)
+                    {
+                        f(inner_it);
+                        ++inner_it;
+                    }
+
+                    if(j < it.range_size - 1  )
+                        for (auto& x: it.M_)
+                            _mm_prefetch(((char*)(&x[j+1])), _MM_HINT_T0);
```
Comment (sithhell) [inline hpx/parallel/util/loop.hpp]: This needs to get handled with a preprocessor conditional. It is a SSE intrinsic and not available on all patforms

File: hpx/parallel/util/loop.hpp [hunk 3]
```diff
@@ -96,23 +261,88 @@ namespace hpx { namespace parallel { namespace util
...
+                    for(std::size_t j = it.idx;j < std::min( (it.range_size ),(it.idx + it.chunk_size) ); ++j)
+                    {
+                        if (tok.was_cancelled())
+                            break;
+                        f(inner_it);
+                        ++inner_it;
+
+                        if(j < it.range_size - 1  )
+                            for (auto& x: it.M_)
+                                _mm_prefetch(((char*)(&x[j+1])), _MM_HINT_T0);
```
Comment (sithhell) [inline hpx/parallel/util/loop.hpp]: Same here.

File: hpx/parallel/util/loop.hpp [hunk 4]
```diff
@@ -68,12 +69,176 @@ namespace hpx { namespace parallel { namespace util
...
+
+            inline std::size_t operator*() const {return idx;}
+        };
+
+
+        //Helper class to initialize prefetching_iterator
+        template<typename T>
+        struct prefetcher_context
+        {
+            std::vector<std::size_t> range;
```
Comment (sithhell) [inline hpx/parallel/util/loop.hpp]: The range of indexes shouldn't be stored as a vector. Could you use boost::irange instead?

File: hpx/parallel/util/loop.hpp [hunk 5]
```diff
@@ -68,12 +69,176 @@ namespace hpx { namespace parallel { namespace util
...
+                return prefetching_iterator<T>(range_size, it_end, chunk_size, range_size, m);
+            }
+
+        };
+
+
+        //function which initialize prefetcher_context
+        template<typename T>
+        prefetcher_context<T> make_prefetcher_context(std::size_t idx_begin, std::size_t idx_end,
+            std::initializer_list< T * > &&l, std::size_t p_factor = 0)
```
Comment (sithhell) [inline hpx/parallel/util/loop.hpp]: I don't like this API of having a pointer + size, couldn't we use something like array_view instead?

**Comment by zkhatami88** [comment]
sithhell: Honestly we are thing about having prefetching as a policy parameter, so it can be used by other algorithms too. Also, current prefetching could be apply easily on transform algorithm too. 

chunk_size used for prefetching here, is determined based on the cache line size. On x86 cache lines are 64 bytes. As far as I know, there is not any command to find out cache line sizes on different architectures. So, I used 64 as a default for determining chunk_size here. 

Instead of using _mm_prefetch, reading the next element in the container can be used for prefetching next chunk_size, but it is more expensive than using _mm_prefetch. Do you have any suggestion?

According to the rage index, first I used boost::irange, but I found it to have a high overhead which prevents scaling. So I changed it to vector, and it scales much better. 

I will try using array_view instead of initializer_list.

**Comment by zkhatami88** [comment]
Can anyone please tell me why my pull request fails?
It builds completely fine, but for "$ docker run -v $PWD:/hpx -w /hpx ${IMAGE_NAME} ./build/bin/inspect --all --output=./build/hpx_inspect_report.html /hpx", it gives "exit code 1".

**Comment by hkaiser** [comment]
@zkhatami88 inspect is unhappy with the code: https://circle-artifacts.com/gh/STEllAR-GROUP/hpx/3961/artifacts/0/tmp/circle-artifacts.sEApRbB/hpx_inspect_report.html

**Comment by sithhell** [comment]
Am 25.05.2016 3:27 nachm. schrieb "Zahra Khatami" <notifications@github.com
>:
>
> sithhell: Honestly we are thing about having prefetching as a policy
parameter, so it can be used by other algorithms too. Also, current
prefetching could be apply easily on transform algorithm too.

It would be really nice to have it in its own headers then.

>
> chunk_size used for prefetching here, is determined based on the cache
line size. On x86 cache lines are 64 bytes. As far as I know, there is not
any command to find out cache line sizes on different architectures. So, I
used 64 as a default for determining chunk_size here.

Turn it into a configurable (through cmake) define).

>
> Instead of using _mm_prefetch, reading the next element in the container
can be used for prefetching next chunk_size, but it is more expensive than
using _mm_prefetch. Do you have any suggestion?

As said. This works only for x86. It will not work on other platforms.
Either ignore it for now or figure out the intrinsics for other platforms
in addition to the fallback you describe.

>
> According to the rage index, first I used boost::irange, but I found it
to have a high overhead which prevents scaling. So I changed it to vector,
and it scales much better.

Having a vector sounds wasteful. That's all I'm saying.

File: hpx/parallel/algorithms/for_each.hpp [hunk 6]
```diff
@@ -71,11 +72,13 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
                         [f, proj](std::size_t /*part_index*/,
                             InIter part_begin, std::size_t part_size) mutable
                         {
+                            typedef typename util::loop_n_iterator_mapping<
+                                InIter>::type iterator_type;
```
Comment (hkaiser) [inline hpx/parallel/algorithms/for_each.hpp]: This could be compined with the `loop_n<>` template.

File: hpx/parallel/util/loop.hpp [hunk 7]
```diff
@@ -48,6 +50,7 @@ namespace hpx { namespace parallel { namespace util
         };
     }
 
+
```
Comment (hkaiser) [inline hpx/parallel/util/loop.hpp]: All of the code related to prefetching should go into a separate file, e.g. `hpx/parallel/util/prefecthing.hpp`

File: hpx/parallel/util/loop.hpp [hunk 8]
```diff
@@ -68,12 +71,194 @@ namespace hpx { namespace parallel { namespace util
...
+        //New random access iterator which is used for prefetching
+        //all containers within lambda functions
+        template<typename T>
+        class prefetching_iterator
+        : public std::iterator<std::random_access_iterator_tag, std::size_t>
+        {
+            public:
+
+            using base_iterator = std::vector<std::size_t>::iterator;
+            std::vector< T * > M_;
```
Comment (hkaiser) [inline hpx/parallel/util/loop.hpp]: The container to be prefetched should be stored as a tuple<> of references, not a vector<> of pointers.

File: hpx/parallel/util/loop.hpp [hunk 9]
```diff
@@ -68,12 +71,194 @@ namespace hpx { namespace parallel { namespace util
...
+
+        //New random access iterator which is used for prefetching
+        //all containers within lambda functions
+        template<typename T>
+        class prefetching_iterator
+        : public std::iterator<std::random_access_iterator_tag, std::size_t>
+        {
+            public:
+
+            using base_iterator = std::vector<std::size_t>::iterator;
```
Comment (hkaiser) [inline hpx/parallel/util/loop.hpp]: Please don't use `using` to declare type aliases, use `typedef` instead. We still support compilers which don't (fully) support `using`.

File: hpx/parallel/util/loop.hpp [hunk 10]
```diff
@@ -68,12 +71,194 @@ namespace hpx { namespace parallel { namespace util
...
+            std::size_t range_size;
+            std::size_t idx;
+
+            explicit prefetching_iterator(std::size_t idx_,
+                base_iterator base_ , std::size_t chunk_size_,
+                std::size_t range_size_, std::vector< T * > const & A)
+            : M_(A), base(base_), chunk_size(chunk_size_),
+                range_size(range_size_),idx(idx_) {}
+
+            using difference_type = typename std::iterator<
```
Comment (hkaiser) [inline hpx/parallel/util/loop.hpp]: Same as above, please use `typedef`.

File: hpx/parallel/util/loop.hpp [hunk 11]
```diff
@@ -68,12 +71,194 @@ namespace hpx { namespace parallel { namespace util
...
+            inline bool operator>=(const prefetching_iterator& rhs) const
+            {
+                return idx >= rhs.idx;
+            }
+            inline bool operator<=(const prefetching_iterator& rhs) const
+            {
+                return idx <= rhs.idx;
+            }
+            inline std::size_t operator*() const {return idx;}
+        };
```
Comment (hkaiser) [inline hpx/parallel/util/loop.hpp]: This iterator implementation misses `operator[]`.

File: hpx/parallel/util/loop.hpp [hunk 12]
```diff
@@ -68,12 +71,194 @@ namespace hpx { namespace parallel { namespace util
...
+        {
+            std::vector<std::size_t> range;
+            std::vector<std::size_t>::iterator it_begin;
+            std::vector<std::size_t>::iterator it_end;
+            std::size_t chunk_size;
+            std::vector< T * > m;
+            std::size_t range_size;
+            explicit prefetcher_context (std::size_t begin, std::size_t end,
+                std::size_t p_factor, std::initializer_list< T * > &&l)
+            {
```
Comment (hkaiser) [inline hpx/parallel/util/loop.hpp]: Wouldn't it be better to expect the list of containers as a variadic template parameter list? I.e. something like:

    auto ctx = prefetcher(begin, end, factor, a, b, c);

where a, b, and c are the containers to prefetch.

File: hpx/parallel/util/loop.hpp [hunk 13]
```diff
@@ -68,12 +71,194 @@ namespace hpx { namespace parallel { namespace util
...
+        template<typename T>
+        struct prefetcher_context
+        {
+            std::vector<std::size_t> range;
+            std::vector<std::size_t>::iterator it_begin;
+            std::vector<std::size_t>::iterator it_end;
+            std::size_t chunk_size;
+            std::vector< T * > m;
+            std::size_t range_size;
+            explicit prefetcher_context (std::size_t begin, std::size_t end,
```
Comment (hkaiser) [inline hpx/parallel/util/loop.hpp]: I think the prefetcher should support arbitrary iterators, and not just integral ranges.

File: hpx/parallel/util/loop.hpp [hunk 14]
```diff
@@ -96,23 +281,98 @@ namespace hpx { namespace parallel { namespace util
...
+                std::size_t cnt = count + 1;
+                for (/**/; cnt != 0; (void) --cnt, ++it)
+                {
+                    std::vector<std::size_t>::iterator inner_it = it.base;
+                    std::size_t j = it.idx;
+                    std::size_t last = it.idx+it.chunk_size;
+
+                    //based on HPX Inspection Report,
+                    //using std::min violates boost min/max
+                    if (it.range_size < last)
```
Comment (hkaiser) [inline hpx/parallel/util/loop.hpp]: What the inspect was trying to tell you is that you shouldn't use `std::min(a, b)` but write it as `(std::min)(a, b)`. This prevents `min()` from being macro expanded (which happens on some platforms, unfortunately).

File: hpx/parallel/util/loop.hpp [hunk 15]
```diff
@@ -96,23 +281,98 @@ namespace hpx { namespace parallel { namespace util
...
+                        last = it.range_size;
+
+                    for(/**/; j < last ; ++j)
+                    {
+                        f(inner_it);
+                        ++inner_it;
+                    }
+                    if(j < it.range_size - 1)
```
Comment (hkaiser) [inline hpx/parallel/util/loop.hpp]: I'd suggest to use prefetching intrinsics here, at least on those platforms where they are available.

File: hpx/parallel/util/loop.hpp [hunk 16]
```diff
@@ -96,23 +281,98 @@ namespace hpx { namespace parallel { namespace util
...
+                                x[j+1];
+                    }
+                }
+                return it;
+            }
+
+        };
     }
 
+    template <typename Iter>
```
Comment (hkaiser) [inline hpx/parallel/util/loop.hpp]: I'd suggest merging this template with the `loop_n<>` above.

File: hpx/parallel/util/loop.hpp (hunk 15 above)
Comment (sithhell) [inline hpx/parallel/util/loop.hpp]: Am 03.06.2016 5:17 nachm. schrieb "Hartmut Kaiser" <notifications@github.com
>:
>
> In hpx/parallel/util/loop.hpp:
>
> > +                {
> > +                    std::vector<std::size_t>::
> iterator inner_it = it.base;
> > +                    std::size_t j = it.idx;
> > +                    std::size_t last = it.idx+it.chunk_size;
> > +
> > +                    //based on HPX Inspection Report,
> > +                    //using std::min violates boost min/max
> > +                    if (it.range_size < last)
> > +                        last = it.range_size;
> > +
> > +                    for(/**/; j < last ; ++j)
> > +                    {
> > +                        f(inner_it);
> > +                        ++inner_it;
> > +                    }
> > +                    if(j < it.range_size - 1)
>
> I'd suggest to use prefetching intrinsics here, at least on those
platforms where they are available.

Absolutely. But they have to be properly guarded.

**Comment by zkhatami88** [comment]
@hkaiser : My tests failed because of:
loop.hpp:108:32: error: no member named 'prefetching_iterator' in namespace 'hpx::parallel::util::detail'
        struct loop_n <detail::prefetching_iterator<Itr, TS...> >

which I have " prefetching_iterator" within "prefetching.hpp" [https://github.com/zkhatami88/hpx/blob/master/hpx/parallel/util/prefetching.hpp] and within "hpx::parallel::util::detail" name space. I have included this in header in loop.hpp, but still it seems it cannot find it here in tests. While it doesnt have any problem when I use it in my tests on my own computer.

Can you please help me to fix this?

**Comment by hkaiser** [comment]
@zkhatami88 I will have a look

File: hpx/parallel/util/loop.hpp [hunk 17]
```diff
@@ -8,17 +8,20 @@
...
+#include <vector>
 
 namespace hpx { namespace parallel { namespace util
 {
     ///////////////////////////////////////////////////////////////////////////
     namespace detail
     {
         ///////////////////////////////////////////////////////////////////////
-        // Helper class to repeatedly call a function starting from a given
+        // Helper class to repeatedly call a function starting from a givena
```
Comment (sithhell) [inline hpx/parallel/util/loop.hpp]: there is a typo here.

File: hpx/parallel/util/prefetching.hpp [hunk 18]
```diff
@@ -0,0 +1,224 @@
...
+        inline typename std::enable_if<I == sizeof...(TS), void>::type
+        prefetch_(hpx::util::tuple<TS...>& t, std::size_t idx)
+        {}
+
+        template<std::size_t I = 0, typename... TS>
+        inline typename std::enable_if<I < sizeof...(TS), void>::type
+        prefetch_(hpx::util::tuple<TS...>& t, std::size_t idx)
+        {
+            _mm_prefetch(((char*)(&hpx::util::get<I>(t)[idx])), _MM_HINT_T0);
+            prefetch_<I + 1, TS...>(t, idx);
```
Comment (sithhell) [inline hpx/parallel/util/prefetching.hpp]: Sequencing instead of recursion would be nice here!

File: loop.hpp [hunk 19]
```diff
@@ -0,0 +1,489 @@
...
+                    if (it.range_size < last)
+                        last = it.range_size;
+
+                    for(/**/; j < last ; ++j)
+                    {
+                        f(inner_it);
+                        ++inner_it;
+                    }
+
+#ifdef SUPPORT_mm_prefetch
```
Comment (sithhell) [inline loop.hpp]: This is never set. It would be nice if this logic would be in prefetcher.hpp. Somthing like this:
```
#if defined(HPX_HAVE_X86_SSE_PREFETCH
template <typename Tuple>
void prefetch(Tuple &)
{
    int sequencer[] = {0, (/* do prefetch magic for x86*/, 0)...};
}
#elif defined(HPX_HAVE_SOME_OTHER_PREFETCH
template <typename Tuple>
void prefetch(Tuple &)
{
    int sequencer[] = {0, (/* do prefetch magic for other*/, 0)...};
}
#else
template <typename Tuple>
void prefetch(Tuple &)
{
    int sequencer[] = {0, (hpx::util::get<I>(t)[j+1], 0)...};
}
#endif
```

**Comment by hkaiser** [comment]
@zkhatami88 FWIW, I'm working on fixing things [here](https://github.com/hkaiser/hpx/tree/zahra_master), feel free to chime in.

**Comment by hkaiser** [comment]
@zkhatami88 I'd be happy to merge this once the conflicts with our master branch are resolved.

**Comment by zkhatami88** [comment]
@hkaiser sure, that's a good idea. Also, the test is failed, I am going to fix it too.

**Comment by sithhell** [comment]
My comments haven't been addressed.

**Comment by zkhatami88** [comment]
@sithhell Thank you so much for your comments. I fixed typo in loop.hpp and in prefetching.hpp, sequencer is implemented to have iteration over pefetching containers in each time step instead of having recursion.

File: hpx/parallel/algorithms/for_each.hpp [hunk 20]
```diff
@@ -21,7 +21,7 @@
 #include <hpx/parallel/algorithms/detail/is_negative.hpp>
 #include <hpx/parallel/util/detail/algorithm_result.hpp>
 #include <hpx/parallel/util/foreach_partitioner.hpp>
-#include <hpx/parallel/util/loop.hpp>
+#include <hpx/parallel/util/prefetching.hpp>
```
Comment (hkaiser) [inline hpx/parallel/algorithms/for_each.hpp]: This change is not correct. We should not pull in the `prefetch.hpp` header into the `for_each.hpp`. If a user wants to use prefetching he/she should explicitly include an appropriate header file. Also, for_each depends on `loop.hpp`, so why did you remove it?

**Comment by zkhatami88** [comment]
@hkaiser loop.hpp is included in the header in prefetching.hpp, so by having prefetching.hpp in the headers in for_each.hpp, loop.hpp  will be included as well.
But my main reason was that in my own test, when I wanted to include prefetching.hpp separately in main.cpp file, there were so many compilation error. After including prefetching.hpp in for_each.hpp, the problem solved. Also, with this way, user should be care of header. He/She can use prefetching or not using prefetching with the same headers.

**Comment by hkaiser** [comment]
@zkhatami88 for_each depends on `loop.hpp`, but not on `prefetching.hpp`, thus it should include the appropriate header. Also, we should avoid relying on implicit #includes as much as possible. Every header file should include the minimal but complete set of headers it depends on. We do this to a) make every header self-contained (which is checked by CircleCI), and b) minimize the compile time for each of the headers (which unfortunately we can't check currently).

If including `prefetching.hpp` in user code leads to compilation errors, those have to be fixed separately, but most likely your user code missed other includes in the first place.

**Comment by zkhatami88** [comment]
@hkaiser oh ok. I am going to fix it and I will push another commit.

File: hpx/parallel/algorithms/for_each.hpp [hunk 21]
```diff
@@ -223,22 +226,15 @@ namespace hpx { namespace parallel { HPX_INLINE_NAMESPACE(v1)
             sequential(ExPolicy, InIter first, InIter last, F && f,
                 Proj && proj)
             {
+                typedef typename util::detail::loop<InIter>::type it_type;
+
                 return util::loop(first, last,
                     [&f, &proj](Iter curr)
                     {
                         f(hpx::util::invoke(proj, *curr));
                     });
```
Comment (zkhatami88) [inline hpx/parallel/algorithms/for_each.hpp]: Test fails because of this line, that:

'/hpx/hpx/parallel/algorithms/for_each.hpp:231:24: error: no viable conversion from returned value of type 'hpx::local_raw_vector_iterator<int, __gnu_cxx::__normal_iterator<int *, std::vector<int, std::allocator<int> > > >' to function return type 'hpx::local_vector_iterator<int>'

In all lambdas in this file, ' hpx::util::invoke(f, hpx::util::invoke(proj, *curr)); ' is used except this line, shouldn't we change 'f(hpx::util::invoke(proj, *curr)); '  to ' hpx::util::invoke(f, hpx::util::invoke(proj, *curr)); '  as well?

Comment (hkaiser) [inline hpx/parallel/algorithms/for_each.hpp]: I'll have a look

**Comment by hkaiser** [comment]
This is superseded by #2228
